Use PHP 8 language features (using Rector)

luchaos opened the pull request.

Jamiras commented on lib/database/user.php:
```diff
@@ -1491,54 +1410,34 @@ function GetDeveloperStatsFull($count, $sortBy, $devFilter = 7)
+        5 => "LastLogin DESC",
+        6 => "Author ASC",
+        default => "Achievements DESC",
+    };
```
These changes will make PHP 8.0 a requirement. (see my question about the upgrade being required in#935). It seems like that needs to happen first, but the title of this PR suggests it's a prerequisite for PHP 8. I guess they both have to go together? 

Also, is this still a draft?

luchaos replied:
> the title of this PR suggests it's a prerequisite for PHP 8

Oh, that's misleading then, sorry.

As mentioned in the other PR, this PR here will use PHP 8 language features and require the application to actually run on PHP 8 whereas https://github.com/RetroAchievements/RAWeb/pull/935 merely takes care of a few deprecation warnings that surface when running PHP 7.4 code with PHP 8.

Now that https://github.com/RetroAchievements/RAWeb/pull/935 is merged we can safely switch to PHP 8 on the server.

After that's done this PR can be merged and deployed.
I'm still finding things to clean up, that's why I haven't yet removed the draft label. Will be ready very soon.

luchaos replied:
PR is ready now - I have to stop looking for things to clean up, it's getting out of hand.

luchaos commented:
Added `smtp` mailer option (Symfony), removed `sendmail` option.

Mailhog can be used locally via docker:
![Screenshot 2022-04-11 at 02 17 10](https://user-images.githubusercontent.com/1280590/162646284-54169089-2383-421f-bbe9-f64a9c20b68f.png)
.

Jamiras commented on src/Shortcode/Shortcode.php:
```diff
@@ -225,18 +225,18 @@ private function autoEmbedYouTube($text)
                 | youtube\.com      # or youtube.com followed by
                   \S*               # Allow anything up to VIDEO_ID,
-                  [^\w\-\s]         # but char before ID is non-ID char.
+                  [^\w\\-\s]        # but char before ID is non-ID char.
```
Was this intentional?

luchaos replied:
More or less. Rector changed that using the `RegexDashEscapeRector` rule.
Interestingly enough PHP's regex dialect does not make much of a difference there, PHPStorm's internal syntax check gave me similar responses in the past, ultimately resulting in me changed quite a few of the other regexes. 

I verified youtube emebs still work as intended via `/test/shortcode.php`.

luchaos commented on lib/util/mail.php:
```diff
@@ -179,19 +195,7 @@ function sendFriendEmail($user, $email, $type, $friend)
-        error_log($email);
-        error_log($emailTitle);
-        error_log($msg);
-        $retVal = true;
```
Great idea 👍  I deemed those `isAtHome` usages superfluous as it always returned false (those checks have been replaced with `APP_ENV` checks a while ago).

I added `mail_log` as the default behaviour which always returns `true`, I think that's a sensible default in any case. 
Would that work for you?

luchaos commented on .gitignore:
```diff
@@ -1,16 +1,9 @@
-/lib/database/releases.php
```
You're absolutely right - same here. Not sure why I opted for that to be honest. Reverted.

luchaos commented on lib/database/activity.php:
```diff
@@ -210,8 +206,6 @@ function postActivity($userIn, $activity, $customMsg, $isalt = null)
             $query .= "(NOW(), $activity, '$user', '$lbID', '$score')";
             break;
-
-        default:
```
Good catch, Rector didn't like `default:` not to be the last case statement. 
Moving it below `case ActivityType::Unknown:` makes Rector happy while preserving logic.

luchaos commented on lib/database/game.php:
```diff
@@ -107,47 +104,21 @@ function getGameMetadataByFlags(
+        1 => "ORDER BY ach.DisplayOrder, ach.ID ASC ",
+        11 => "ORDER BY ach.DisplayOrder DESC, ach.ID DESC ",
+        2 => "ORDER BY NumAwarded, ach.ID ASC ",
+        12 => "ORDER BY NumAwarded DESC, ach.ID DESC ",
```
Good question - I spent some time with that one. Most of the options are self-explaining imo. 

I deemed `DisplayOrder` generic enough to keep the reference to `by the developer`.

Restored the comment on `3 and 13`, deleted that one in the heat.

luchaos commented on lib/database/game.php:
```diff
@@ -107,47 +104,21 @@ function getGameMetadataByFlags(
+        14 => "ORDER BY ach.Points DESC, ach.ID DESC ",
+        5 => "ORDER BY ach.Title, ach.ID ASC ",
+        15 => "ORDER BY ach.Title DESC, ach.ID DESC ",
+        default => "ORDER BY ach.DisplayOrder, ach.ID ASC ",
```
Tried to keep the order there, wasn't too happy with the duplicated option either.
Added `$sortBy ??= 1;` and removed the default, should do the trick.

luchaos replied:
Now that didn't work as intended at all. Moved case `1` to `default` instead.

luchaos commented on lib/database/leaderboard.php:
```diff
@@ -525,25 +504,26 @@ function GetLeaderboardData($lbID, $user, $numToFetch, $offset, $friendsOnly, $n
                         unset($db_entry['UserRank']);
+                        // @phpstan-ignore-next-line
                         $db_entry['Index'] = (int) $db_entry['UserIndex'];
+                        // @phpstan-ignore-next-line
```
Very good question! I wasn't too sure either, didn't make much sense to me either. 

It seems to happen because the same variable name with similar keys is used above. To fix this it should be refactored into a temporary variable to only hold the desired keys without the need to `unset` anything.

I try to refrain from using static analysis or style ignores. Cleaning that up will take some more attention to the function's logic, hence why I didn't take care of it yet (this function is far too long, too).

luchaos commented on lib/database/game.php:
```diff
@@ -383,10 +354,6 @@ function getGamesListByDev($dev, $consoleID, &$dataOut, $sortBy, $ticketsFlag =
     }
 
     switch ($sortBy) {
```
Transforming `switch` to `match` was mostly done by Rector. 

It will however not do that if cases contain complex logic (in this case `if/else`), `match` doesn't support that (which is good as prevents complexity within cases).

Rephrased it to ternary operators and Rector converted it to `match`. (Did that somewhere else, too, iirc)
Additionally, added case `1`, `$sortBy ??= 1;`, and removed the default - similar to the other match statement above.

luchaos replied:
See https://github.com/RetroAchievements/RAWeb/pull/911#discussion_r847636111

luchaos commented on lib/database/setrequest.php:
```diff
@@ -78,15 +77,15 @@ function getUserRequestsInformation($user, $list, $gameID = -1)
-        //If the game does not have achievements then it couns as a legit request
-        if (count(getAchievementIDs($request['GameID'])['AchievementIDs']) == 0) {
+        // If the game does not have achievements then it couns as a legit request
+        if ((is_countable(getAchievementIDs($request['GameID'])['AchievementIDs']) ? count(getAchievementIDs($request['GameID'])['AchievementIDs']) : 0) == 0) {
```
Oh, seems I didn't catch them all, thanks!

luchaos commented on lib/database/user.php:
```diff
@@ -2,36 +2,32 @@
     $emailCookie = rand_string(16);
-    // $expiry = date('Y-m-d', time() + 60 * 60 * 24 * 7);
-    $expiry = time() + 60 * 60 * 24 * 7;
+    $expiry = date('Y-m-d', time() + 60 * 60 * 24 * 7);
```
Interestingly enough, that column's type has already been set to `date` which actually does not allow to be set with an integer value (unix timestamp). 

At least that's true for MySQL 8. Maybe it was allowed in MySQL 5.7 with more lenient SQL modes which was used when I started working with this codebase. It seems I/someone found that issue before but never went through with it, leaving that line commented out. 

It solves a fairly long standing issue which causes all values in the production database to be `0000-00-00` for `Expires`.

I verified that it works properly now locally. Just to be sure, did you get different results?

Jamiras replied:
> did you get different results?

I didn't try. I was just looking at the code changes.

luchaos commented on lib/util/log.php:
```diff
@@ -1,29 +1,15 @@
-    // }
+    if (filter_var(getenv('APP_DEBUG'), FILTER_VALIDATE_BOOLEAN, FILTER_NULL_ON_FAILURE)) {
+        throw new Exception($error);
+    }
```
Good catch. Moved.

luchaos commented on public/gameInfo.php:
```diff
@@ -805,7 +794,7 @@ function submitSetRequest(user, gameID) {
                 if ($numAchievements > 0) {
                     echo "<b>Authors:</b> ";
-                    $numItems = count($authorInfo);
+                    $numItems = count((array) $authorInfo);
```
I think that was a PHPStan error - it's initialized with `null` but assumed to be iterable/countable from then on.
I changed the initial value to `[]` instead, that's even cleaner seeing it's both counted and iterated here.

luchaos commented on lib/database/forum.php:
```diff
@@ -433,35 +410,25 @@ function generateGameForumTopic($user, $gameID, &$forumTopicID)
+    $forumID = match ($consoleID) {
+        // Mega Drive
+        1 => 10,
+        // SENS
```
unforgivable 🤦  thanks 😊 
> Save `SNES` to dictionary

Jamiras commented:
I'm not sure if I missed it, but is rector part of the push hook now?

luchaos commented:
> I'm not sure if I missed it, but is rector part of the push hook now?
I wouldn't add Rector to a CI pipeline or git hook. Not only can it take quite long to run; I think it's meant to be run on demand. At least that's how I use it atm. Helps with simplifying code that is not yet using the most recent language features, but it's not necessary to do so.

I'll research how other projects integrate it into their workflows.
